databricks-jobs/dbjobs.py

```diff
@@ -0,0 +1,213 @@
+        jsonfile: Path to the file containing json job definition
+
+    """
+    with open(jsonfile) as f:
+        job_settings_map = json.load(f)
+    data = {}
+    data['job_id'] = jobid
+    data['new_settings'] = job_settings_map
+    json_string = json.dumps(data)
+    print('Attempting to update jobid {} with settings {}'.format(jobid, json_string))
+    jobs_rpc('reset', json_string)
+
+def create_job(jsonfile):
+    """ Creates a new job
+
+    Args:
+        jsonfile: Path to the file containing json job definition
+
+    Returns:
+        jobid: The new job's id number
+
+    """
+    with open(jsonfile) as f:
+        job_settings_map = json.load(f)
+    job_settings_string = json.dumps(job_settings_map)
+    response_json = jobs_rpc('create', job_settings_string)
+    return response_json['job_id']
+
+def stop_job(job_id):
+    """ Stops all runs of a job, polls until stopped
+
+    Args:
+        job_id: The id of the job to stop all runs for
+
+    """
+    data = {}
+    data['active_only'] = 'true'
+    data['job_id'] = job_id
+    # 0 specifies a maximum limit of runs to return
+    data['limit'] = 0
+    json_string = json.dumps(data)
+    response_json = runs_rpc('list', json_string)
+
+    if 'runs' not in response_json:
+        print('job_id {} had no runs'.format(job_id))
+        return
+
+    runids = []
+    for run in response_json['runs']:
+        runids.append(run['run_id'])
+
+    for runid in runids:
+        print('Asynchronously cancelling run_id {} for job_id {}'.format(runid, job_id))
+        request = {}
+        request['run_id'] = runid
+        request_string = json.dumps(request)
+
+        response = requests.post(
+            RUNS_BASE_URL + 'cancel',
+            headers = {"Authorization": "Basic " + base64.standard_b64encode("token:" + DATABRICKS_TOKEN)},
+            data = request_string
+            )
+
+        response_dict = response.json()
+        # If response dict is non empty, print contents which contain error
+        if response_dict:
+            print('Cancel job call has likely failed. Response was nonempty: {}'.format(str(response_dict)))
+            return
+
+        print('Poll until run_id {} is stopped'.format(runid))
+        try:
+            poll(
+                lambda: poll_job_completed(request_string) == True,
+                step = 10,
+                timeout = 300
+                )
+            print('Run_id {} is in a terminal state'.format(runid))
+        except TimeoutException:
+            print('Timed out waiting for run_id {} to terminate'.format(runid))
+
+
+def poll_job_completed(request_string):
+    """ Helper function for polling until the job is in a terminal state """
+    response_json = runs_rpc('get', request_string)
+    if (response_json['state']['life_cycle_state'] == "TERMINATED" or
+        response_json['state']['life_cycle_state'] == "SKIPPED" or
+        response_json['state']['life_cycle_state'] == "INTERNAL_ERROR"):
```

Comment:
Is `INTERNAL_ERROR` a state we should worry about?

Reply:
not when attempting to stop a job.  If a job errors out we should know beforehand either by email, or build additional monitoring of state.